src/main/java/teammates/ui/controller/InstructorStudentListRemindAction.java

```diff
@@ -0,0 +1,137 @@
+import teammates.common.util.StatusMessage;
+import teammates.common.util.StatusMessageColor;
+import teammates.common.util.StringHelper;
+
+/**
+ * Action: remind instructor or student to register for a course by sending reminder emails.
+ */
+public class InstructorStudentListRemindAction extends Action {
+
+    @Override
+    public ActionResult execute() throws EntityDoesNotExistException {
+
+        String courseId = getRequestParamValue(Const.ParamsNames.COURSE_ID);
+        Assumption.assertPostParamNotNull(Const.ParamsNames.COURSE_ID, courseId);
+
+        CourseAttributes course = logic.getCourse(courseId);
+        if (course == null) {
+            throw new EntityDoesNotExistException("Course with ID " + courseId + " does not exist!");
+        }
+
+        String studentEmail = getRequestParamValue(Const.ParamsNames.STUDENT_EMAIL);
+        String instructorEmail = getRequestParamValue(Const.ParamsNames.INSTRUCTOR_EMAIL);
+
+        InstructorAttributes instructor = logic.getInstructorForGoogleId(courseId, account.googleId);
+        boolean isSendingToStudent = studentEmail != null;
+        boolean isSendingToInstructor = instructorEmail != null;
+        if (isSendingToStudent) {
+            gateKeeper.verifyAccessible(
+                    instructor, course, Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_STUDENT);
+        } else if (isSendingToInstructor) {
+            gateKeeper.verifyAccessible(
+                    instructor, course, Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_INSTRUCTOR);
+        } else {
+            // this is sending registration emails to all students in the course and we will check if the instructor
+            // canmodifystudent for course level since for modifystudent privilege there is only course level setting for now
+            gateKeeper.verifyAccessible(
+                    instructor, course, Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_STUDENT);
+        }
+
+        /* Process sending emails and setup status to be shown to user and admin */
+        Map<String, JoinEmailData> emailDataMap = new TreeMap<>();
+
+        String redirectUrl = "";
+        if (isSendingToStudent) {
+            taskQueuer.scheduleCourseRegistrationInviteToStudent(courseId, studentEmail, false);
+            StudentAttributes studentData = logic.getStudentForEmail(courseId, studentEmail);
+            if (studentData == null) {
+                throw new EntityDoesNotExistException("Student with email " + studentEmail + " does not exist "
+                        + "in course " + courseId + "!");
+            }
+            emailDataMap.put(studentEmail,
+                    new JoinEmailData(studentData.getName(), extractStudentRegistrationKey(studentData)));
+
+            statusToUser.add(new StatusMessage(Const.StatusMessages.COURSE_REMINDER_SENT_TO + studentEmail,
+                    StatusMessageColor.SUCCESS));
+            redirectUrl = Const.ActionURIs.INSTRUCTOR_STUDENT_LIST_PAGE;
```

Comment:
This file is the same as InstructorCourseRemindAction.java except for line 65 and line 95. That is quite wasteful; can you think of a way use only one class and achieve both features?

Reply:
Yes it is. 
I can make a single class for reminding students and use it for both InstructorCourseDetails and InstructorStudentList by adding an if else check, but doing so might require changing code for both pages (deletion and addition of a series of files including studentList.tag, Data and Action files). Will that be okay?

Reply:
If that doesn't leave too many lines of code changes, I think that's fine :) be mindful of inheritance though because many classes inherit from Action so might have to make changes to those files as well.

Reply:
I will try to be careful of the changes I make from here on out :)

In addition to what I said above, would it be appropriate to simply use requestURL, isolate the string and check the page (if it is instructorStudentList or instructoCourseDetails)? I have noticed that it already has the link to current page as "/page/instructorCourseRemind?...." so taking the string from the last '/' to the first '?' might solve the issue as well as minimize the changes to the code files.

Reply:
Thank you so much for all the help! :)
I have a proposed solution that uses request.getHeader("referer") to get the current (i.e. page that redirects) as a string in Action. I can compare that with the ActionUri in the concerned page and redirect accordingly. Do you think that this is the right way to go? It also involves trimming and comparing the acquired current page string in the InstructorCourseRemindAction.

Reply:
The solution sounds promising, give it a try!